src/diagnostics.rs

```diff
@@ -0,0 +1,324 @@
+use deno_ast::SourceRange;
+use deno_ast::SourceRangedForSpanned;
+use deno_graph::type_tracer::EsmModuleSymbol;
+use deno_graph::type_tracer::ExportDeclRef;
+use deno_graph::type_tracer::SymbolId;
+use deno_graph::type_tracer::SymbolNodeRef;
+use deno_graph::type_tracer::UniqueSymbolId;
+
+use std::collections::HashSet;
+use std::fmt;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DocDiagnosticKind {
+  MissingJsDoc,
+  MissingTypeRef,
+  PrivateTypeRef,
+}
+
+impl std::fmt::Display for DocDiagnosticKind {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    match self {
+      DocDiagnosticKind::MissingJsDoc => {
+        f.write_str("Missing JS documentation comment.")
+      }
+      DocDiagnosticKind::MissingTypeRef => {
+        f.write_str("Missing explicit type.")
+      }
+      DocDiagnosticKind::PrivateTypeRef => {
+        f.write_str("Type is not exported, but referenced by an exported type.")
+      }
+    }
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct DocDiagnostic {
+  pub location: Location,
+  pub kind: DocDiagnosticKind,
+}
+
+#[derive(Default)]
+pub struct DiagnosticsCollector {
+  seen_private_types_in_public: HashSet<UniqueSymbolId>,
+  seen_jsdoc_missing: HashSet<Location>,
+  seen_missing_type_refs: HashSet<Location>,
+  diagnostics: Vec<DocDiagnostic>,
+}
+
+impl DiagnosticsCollector {
+  pub fn add_private_type_in_public(
+    &mut self,
+    module: &EsmModuleSymbol,
+    symbol_id: SymbolId,
+    node_ref: SymbolNodeRef,
+  ) {
+    let unique_id = UniqueSymbolId {
+      module_id: module.module_id(),
+      symbol_id,
+    };
+    if self.seen_private_types_in_public.insert(unique_id) {
+      let ident_range = get_ident_range_of_node(node_ref);
+      self.diagnostics.push(DocDiagnostic {
+        location: get_location(module.source(), ident_range.start),
+        kind: DocDiagnosticKind::PrivateTypeRef,
+      })
+    }
+  }
+
+  pub fn take_diagnostics(&mut self) -> Vec<DocDiagnostic> {
+    let inner = std::mem::take(&mut self.diagnostics);
+    *self = Default::default(); // reset
+    inner
+  }
+
+  pub fn analyze_doc_nodes(&mut self, doc_nodes: &[DocNode]) {
+    DiagnosticDocNodeVisitor { diagnostics: self }.visit_doc_nodes(doc_nodes)
+  }
+
+  fn add_missing_js_doc(&mut self, location: &Location) {
+    if self.seen_jsdoc_missing.insert(location.clone()) {
+      self.diagnostics.push(DocDiagnostic {
+        location: location.clone(),
+        kind: DocDiagnosticKind::MissingJsDoc,
+      })
+    }
+  }
+
+  fn add_missing_type_ref(&mut self, location: &Location) {
+    if self.seen_missing_type_refs.insert(location.clone()) {
+      self.diagnostics.push(DocDiagnostic {
+        location: location.clone(),
+        kind: DocDiagnosticKind::MissingTypeRef,
+      })
+    }
+  }
+}
+
+struct DiagnosticDocNodeVisitor<'a> {
+  diagnostics: &'a mut DiagnosticsCollector,
+}
+
+impl<'a> DiagnosticDocNodeVisitor<'a> {
+  pub fn visit_doc_nodes(&mut self, doc_nodes: &[DocNode]) {
```

Comment:
So it's pretty much a lint rule? 😅

Reply:
Yes, but much smarter than something like https://github.com/denoland/deno_lint/pull/972

Reply:
Should we close that issue because of this feature?

Reply:
Yeah, I'll have the CLI feature PR close it.